core/trino-main/src/test/java/io/trino/operator/exchange/TestLocalExchange.java

```diff
@@ -304,6 +315,82 @@ public void testRandom()
+                nodePartitioningManager,
+                SESSION,
+                3,
+                SCALED_WRITER_DISTRIBUTION,
+                ImmutableList.of(),
+                TYPES,
+                Optional.empty(),
+                DataSize.ofBytes(retainedSizeOfPages(4)),
+                TYPE_OPERATOR_FACTORY,
+                physicalWrittenBytes::get,
+                DataSize.ofBytes(retainedSizeOfPages(2)));
+
+        run(localExchange, exchange -> {
+            assertEquals(exchange.getBufferCount(), 3);
+            assertExchangeTotalBufferedBytes(exchange, 0);
+
+            LocalExchangeSinkFactory sinkFactory = exchange.createSinkFactory();
+            sinkFactory.noMoreSinkFactories();
+            LocalExchangeSink sink = sinkFactory.createSink();
+            assertSinkCanWrite(sink);
+            sinkFactory.close();
+
+            LocalExchangeSource sourceA = exchange.getSource(0);
+            assertSource(sourceA, 0);
+
+            LocalExchangeSource sourceB = exchange.getSource(1);
+            assertSource(sourceB, 0);
+
+            LocalExchangeSource sourceC = exchange.getSource(2);
+            assertSource(sourceC, 0);
+
+            sink.addPage(createPage(0));
+            sink.addPage(createPage(0));
+            assertEquals(sourceA.getBufferInfo().getBufferedPages(), 2);
+            assertEquals(sourceB.getBufferInfo().getBufferedPages(), 0);
+            assertEquals(sourceC.getBufferInfo().getBufferedPages(), 0);
+
+            // writer min file and buffered data size limits are breached, so we should see pages in sourceB
+            physicalWrittenBytes.set(retainedSizeOfPages(2));
+            sink.addPage(createPage(0));
+            assertEquals(sourceA.getBufferInfo().getBufferedPages(), 2);
+            assertEquals(sourceB.getBufferInfo().getBufferedPages(), 1);
+            assertEquals(sourceC.getBufferInfo().getBufferedPages(), 0);
+
+            // no limit is breached, so we should see round-robin distribution across sourceA and sourceB
+            physicalWrittenBytes.set(retainedSizeOfPages(3));
+            sink.addPage(createPage(0));
+            assertEquals(sourceA.getBufferInfo().getBufferedPages(), 3);
+            assertEquals(sourceB.getBufferInfo().getBufferedPages(), 1);
+            assertEquals(sourceC.getBufferInfo().getBufferedPages(), 0);
+
+            // writer min file and buffered data size limits are breached again, but according to
+            // round-robin sourceB should receive a page
+            physicalWrittenBytes.set(retainedSizeOfPages(4));
+            sink.addPage(createPage(0));
+            assertEquals(sourceA.getBufferInfo().getBufferedPages(), 3);
+            assertEquals(sourceB.getBufferInfo().getBufferedPages(), 2);
+            assertEquals(sourceC.getBufferInfo().getBufferedPages(), 0);
+
+            assertSinkWriteBlocked(sink);
+            assertRemovePage(sourceA, createPage(0));
+
```

Comment:
We should also test that there is no scaling when only writer min size or only buffered data size limits are breached